yarn-project/sequencer-client/src/sequencer/public_processor.ts

```diff
@@ -116,15 +119,19 @@ export class PublicProcessor {
     for (const tx of txs) {
       this.log(`Processing tx ${await tx.getTxHash()}`);
       try {
+        await this.publicContractsDB.addNewContracts(tx);
         result.push(await this.processTx(tx));
       } catch (err) {
         this.log.warn(`Error processing tx ${await tx.getTxHash()}: ${err}`);
         failed.push({
           tx,
           error: err instanceof Error ? err : new Error('Unknown error'),
         });
+      } finally {
+        await this.publicContractsDB.clearTxContracts();
```

Comment:
As @spalladino says, ideally we would leave contracts from successfully executed transactions in the cache for later transactions to call. Which I don't think is that difficult given what you have done. And and e2e test to demonstrate it would be great!